src/plugins/enzymes/src/FindEnzymesDialog.cpp

```diff
@@ -771,6 +783,211 @@ bool EnzymeTreeItem::operator<(const QTreeWidgetItem& other) const {
+                                                         (reverseNOut != 0 && reverseNShift.first == Ns::Left),
+                                                         (reverseNIn != 0 && reverseNShift.first == Ns::Left));
+         reverseTooltipElements = generateTooltipElements(reverseNOut, reverseNIn, false, reverseNShift.first,
+                                                         (forwardNOut != 0 && forwardNShift.first == Ns::Left),
+                                                         (forwardNIn != 0 && forwardNShift.first == Ns::Left));
+    } else {
+        if (forwardNShift.first != Ns::No) {
+            forwardTooltipElements = generateTooltipElements(0, forwardNShift.second, true, forwardNShift.first,
+                                                            false,
+                                                            (reverseNShift.second != 0 && reverseNShift.first == Ns::Left));
+        }
+        if (reverseNShift.first != Ns::No) {
+            reverseTooltipElements = generateTooltipElements(0, reverseNShift.second, false, reverseNShift.first,
+                                                            false,
+                                                            (forwardNShift.second != 0 && forwardNShift.first == Ns::Left));
+        }
+    }
+    // generates main part (which contains enzyme)
+    auto generateMainPart = [enzymeSize](const QByteArray& seq, int cut, bool forward) -> QString {
+        QString result;
+        auto append2Result = [&result, forward](const QString& add) {
+            if (forward) {
+                result += add;
+            } else {
+                result.insert(0, add);
+            }
+        };
+        auto removeSpaceFromResult = [&result, forward]() {
+            if (forward) {
+                result = result.left(result.size() - TOOLTIP_SPACE.size());
+            } else {
+                result = result.right(result.size() - TOOLTIP_SPACE.size());
+            }
+        };
+        append2Result(TOOLTIP_SPACE);
+        for (int i = 0; i < enzymeSize; i++) {
+            if (i == cut) {
+                removeSpaceFromResult();
+                append2Result(forward ? TOOLTIP_FORWARD_MARKER : TOOLTIP_REVERSE_MARKER);
+            }
+            append2Result(QString(seq.at(i)));
+            append2Result(TOOLTIP_SPACE);
+        }
+        if (seq.size() == cut) {
+            removeSpaceFromResult();
+            append2Result(forward ? TOOLTIP_FORWARD_MARKER : TOOLTIP_REVERSE_MARKER);
+        }
+        return result;
+    };
+    QString forwardMainPart = generateMainPart(enzyme->seq, enzyme->cutDirect, true);
+    QString reverseMainPart = generateMainPart(seqComplement, enzyme->cutComplement, false);
+
+    // Join parts to a single tooltip
+    auto generateTooltip = [](Ns type, const QStringList& elements, const QString& mainPart) -> QString {
+        QString result;
+        auto appendElements = [&elements]() -> QString {
```

Comment:
nit: join will sound better, append usually mean you append to the current (some) object in the context.

Btw: do I understand correctly that you compute the joined value in any case? In this case why don't you just compute it as variable and append later to the correct part: why all this complexity with the inner function - the code can be simpler and more straightforward (readable)

Reply:
OK

> Btw

I'm not sure how to answer this question, but he main point here that I have thought that this issue would be much easier that it's turned out. The main complexity here is that the direct enzyme depends on the reverse-complement (and vice-versa), so it is very complicated to calculate, for example, the beginning part of the direct enzyme until you know how the whole reverse enzyme looks like. This is why I need so mana pre-calculation functions and join all calculated stuff in the very end. 

You may check the enzyme file which I use in GUI test - there are a lot of different cases which calculates differently, it was very complicated to handle them all.